rfcs/0014-wal_on_disk_format.md

```diff
@@ -0,0 +1,293 @@
+
+<!-- TOC --><a name="operations"></a>
+## Operations
+
+<!-- TOC --><a name="performance-cost"></a>
+### Performance & Cost
+
+<!-- Describe performance and cost implications of this change. -->
+
+- **Latency (reads/writes/compactions):** 
+The latency for writes with durability guarantee mainly depends on the configured size of the encoded WAL object 
+(see `max_wal_bytes_size`) that is used to decide when the in-memory WAL data structure is flushed to object storage.
+Since the new persistence format can batch more records than the currently used SST format given the size of the 
+encoded WAL object, the latency might be higher.
+The latency for writes without durability guarantee might be lower if the in-memory data structure is a simple FIFO
+data structure as recommended in this RFC, since writes to the in-memory WAL are faster.
+A WAL object is read for recovery.
+If we define the latency of recovery as the time until the in-memory state is re-established then the new persistence 
+format might decrease the latency of recovery because more records can be read from object storage given an encoded size 
+of the WAL object.
+The laterncy of compaction is not affected by the new persistence format.
+- **Throughput (reads/writes/compactions):**
+The throughput of writes with durability guarantee might be higher since the new persistence format has less space 
+overhead without indices and filters. 
+Thus, with the new format slateDB can write more WAL records per time unit to object storage compared to the current 
+format given the same encoded size of the WAL object.
+Additionally, the new persistence format allows to recover more records per time unit.
+The throughput of compaction is not affected by the new persistence format.
+- **Object-store request (GET/LIST/PUT) and cost profile:**
+The new persistence format allows to store more records per PUT to and to get more records per GET from object storage. 
+- **Space, read, and write amplification:**
+All three amplifications are reduced. 
+
+<!-- TOC --><a name="observability"></a>
+### Observability
+
+<!-- Describe any operational changes required to support this change. -->
+
+- Configuration changes: 
+	- introduce configuration `wal_flush_size: usize` to specify what encoded size in bytes of the WAL triggers a flush to object storage. 
+	The limit is not strict. The flush is triggered as soon as a write to the WAL exceeds the limit.  
+	- min value is `0`: the WAL is flushed at each write
+	- max value is `4 GiB` (i.e., `4,294,967,295` bytes)
+- New components/services: No
+- Metrics: No
+- Logging: No
+
+<!-- TOC --><a name="compatibility"></a>
+### Compatibility
+
+<!-- Describe compatibility considerations with existing versions of SlateDB. -->
+
+- Existing data on object storage / on-disk formats
+	- To be determined after agreement on the format
+- Existing public APIs (including bindings): No
+- Rolling upgrades / mixed-version behavior (if applicable)
```

Comment:
curious if you have any ideas on how we can do rolling upgrades here, and in general if it makes sense to have shared metadata across the different types of slatedb data files so we can figure out which type it is with a shared metadata schema (e.g. [version | type] is the shared metadata)